src/atem.ts

```diff
@@ -554,4 +605,104 @@ export class Atem extends EventEmitter {
 	}
+
+	/**
+	 * Helper method used by listVisibleInputs.
+	 * This got broken out into its own method because
+	 * it gets called multiple times, and gets called in a loop.
+	 * Breaking it out made listVisibleInputs much easier to read.
+	 */
+	private _calcActiveMeInputs (mode: 'program' | 'preview', meId: number): number[] {
+		const inputs = new Set<number>()
+		const meRef = this.state.video.getMe(meId)
+
+		if (mode === 'preview') {
+			if (meRef.transitionProperties.selection & 1) {
+				inputs.add(meRef.previewInput)
+			}
+		} else {
+			inputs.add(meRef.programInput)
+		}
+
+		// Upstream Keyers
+		Object.values(meRef.upstreamKeyers).filter(usk => {
+			// Pretty gross bitwise operations in this next line, be warned.
+			// TODO(Lange): This feels fragile but I'm unsure how to make it more robust.
+			const isPartOfTransition = meRef.transitionProperties.selection & (2 ** (usk.upstreamKeyerId + 1))
+			if (mode === 'program') {
+				if (meRef.inTransition) {
+					return usk.onAir || isPartOfTransition
+				}
+
+				return usk.onAir
+			}
+
+			return isPartOfTransition
+		}).forEach(usk => {
+			inputs.add(usk.fillSource)
+
+			// This is the only USK type that actually uses the cutSource.
+			if (usk.mixEffectKeyType === Enums.MixEffectKeyType.Luma) {
+				inputs.add(usk.cutSource)
+			}
+		})
+
+		// DSKs only show up on ME 1,
+		// so we only add them if that's the ME we are currently processing.
+		// TODO: Is this still true for the new Constellation ATEMs?
+		if (meId === 0) {
+			Object.values(this.state.video.downstreamKeyers).filter(dsk => {
+				if (mode === 'program') {
+					return dsk.onAir || dsk.inTransition
+				}
+
+				if (!dsk.properties) {
+					// Data isn't hydrated yet, we'll get 'em next time.
+					return
+				}
+
+				return dsk.properties.tie && !dsk.onAir
+			}).forEach(dsk => {
+				inputs.add(dsk.sources.fillSource)
+				inputs.add(dsk.sources.cutSource)
+			})
+		}
+
+		// Compute what sources are currently participating in a transition.
+		// We only care about this for PGM.
+		if (meRef.inTransition && mode === 'program') {
+			if (meRef.transitionProperties.selection & 1) {
+				inputs.add(meRef.previewInput)
+			}
+
+			// From here, what inputs are participating in the transition depends
+			// on the transition style being used, so we handle each separately.
+			switch (meRef.transitionProperties.style) {
+				case Enums.TransitionStyle.DIP:
+					inputs.add(meRef.transitionSettings.dip.input)
+					break
+				case Enums.TransitionStyle.DVE:
+					inputs.add(meRef.transitionSettings.DVE.fillSource)
+					if (meRef.transitionSettings.DVE.enableKey) {
+						inputs.add(meRef.transitionSettings.DVE.keySource)
+					}
+					break
+				case Enums.TransitionStyle.WIPE:
+					if (meRef.transitionSettings.wipe.borderWidth > 0) {
+						inputs.add(meRef.transitionSettings.wipe.borderInput)
+					}
+					break
+				case Enums.TransitionStyle.STING:
+					inputs.add(meRef.transitionSettings.stinger.source)
```

Comment:
These ids need translating, as the `source` is just an index of the media player https://github.com/LibAtem/LibAtem/blob/master/LibAtem/Common/StingerSource.cs